src/rules.jl

```diff
@@ -211,14 +216,14 @@ _abs_deriv(x) = signbit(x) ? -one(x) : one(x)
 @define_diffrule NaNMath.sqrt(x)   = :(  inv(2 * NaNMath.sqrt($x))                  )
 @define_diffrule NaNMath.sin(x)    = :(  NaNMath.cos($x)                            )
 @define_diffrule NaNMath.cos(x)    = :( -NaNMath.sin($x)                            )
-@define_diffrule NaNMath.tan(x)    = :(  1 + NaNMath.pow(NaNMath.tan($x), 2)        )
-@define_diffrule NaNMath.asin(x)   = :(  inv(NaNMath.sqrt(1 - NaNMath.pow($x, 2)))  )
-@define_diffrule NaNMath.acos(x)   = :(  -inv(NaNMath.sqrt(1 - NaNMath.pow($x, 2))) )
-@define_diffrule NaNMath.acosh(x)  = :(  inv(NaNMath.sqrt(NaNMath.pow($x, 2) - 1))  )
-@define_diffrule NaNMath.atanh(x)  = :(  inv(1 - NaNMath.pow($x, 2))                )
+@define_diffrule NaNMath.tan(x)    = :(  1 + NaNMath.pow(NaNMath.tan($x), oftype($x, 2))        )
+@define_diffrule NaNMath.asin(x)   = :(  inv(NaNMath.sqrt(1 - NaNMath.pow($x, oftype($x, 2))))  )
+@define_diffrule NaNMath.acos(x)   = :(  -inv(NaNMath.sqrt(1 - NaNMath.pow($x, oftype($x, 2)))) )
+@define_diffrule NaNMath.acosh(x)  = :(  inv(NaNMath.sqrt(NaNMath.pow($x, oftype($x, 2)) - 1))  )
+@define_diffrule NaNMath.atanh(x)  = :(  inv(1 - NaNMath.pow($x, oftype($x, 2)))                )
 @define_diffrule NaNMath.log(x)    = :(  inv($x)                                    )
-@define_diffrule NaNMath.log2(x)   = :(  inv($x) / NaNMath.log(2)                   )
-@define_diffrule NaNMath.log10(x)  = :(  inv($x) / NaNMath.log(10)                  )
+@define_diffrule NaNMath.log2(x)   = :(  inv($x) / NaNMath.log(oftype($x, 2))       )
+@define_diffrule NaNMath.log10(x)  = :(  inv($x) / NaNMath.log(oftype($x, 10))      )
```

Comment:
Couldn't we just use
```suggestion
@define_diffrule NaNMath.log2(x)   = :(  inv($logtwo * $x)  )
@define_diffrule NaNMath.log10(x)  = :(  inv($logten * $x)  )
```
or something similar here?